File: tests/src/commonTest/kotlin/primitives/ReturnPrimitive.kt
package primitives

import withLib
import kotlin.test.Test
import kotlin.test.assertEquals

class ReturnPrimitive {

    @Test
    fun returnVoid() = withLib {
        assertEquals(Unit, natives.test.returnVoid())
    }

    @Test
    fun returnChar() = withLib {
        assertEquals('a', natives.test.returnChar())
    }

    @Test
    fun returnBoolean() = withLib {
        assertEquals(true, natives.test.returnBoolean())
    }

    @Test
    fun returnByte() = withLib {
        assertEquals(99.toByte(), natives.test.returnByte())
    }

    @Test
    fun returnShort() = withLib {
        assertEquals(99.toShort(), natives.test.returnShort())
    }

    @Test
    fun returnInt() = withLib {
        assertEquals(99, natives.test.returnInt())
    }

    @Test
    fun returnLong() = withLib {
        assertEquals(9223372036854775805L, natives.test.returnLong())
    }

    @Test
    fun returnFloat() = withLib {
        assertEquals(99f, natives.test.returnFloat())
    }

    @Test
    fun returnDouble() = withLib {
        assertEquals(99.0, natives.test.returnDouble())
    }

    @Test
    fun returnStringLiteral() = withLib {
        assertEquals("test string", natives.test.returnStringLiteral())
    }

    @Test
    fun returnString() = withLib {
        assertEquals("test string", natives.test.returnString())
    }
}

File: tests/natives/test/src/primitives/return_primitive.c
#include <api.h>
#include <stdlib.h>
#include <string.h>

void returnVoid() {
}

KChar returnChar() {
    return 'a';
}

KBoolean returnBoolean() {
    return true;
}

KByte returnByte() {
    return 99;
}

KShort returnShort() {
    return 99;
}

KInt returnInt() {
    return 99;
}

KLong returnLong() {
    return 9223372036854775805;
}

KFloat returnFloat() {
    return 99;
}

KDouble returnDouble() {
    return 99.0;
}

KString returnStringLiteral() {
    return makeKString("test string", 11);
}

KString returnString() {
    char* str = malloc(100);
    strcpy(str, "test string");
    return makeKString(str, 11);
}

MyEnum returnEnum() {
    return MyEnum_CASE2;
}

MyStruct returnStruct() {
    return (MyStruct){ 1, 2, 3, 4 };
}